### ISSUE TITLE
Null check in validator is not working

### ISSUE THREAD

**Issue by jasonycw** [JeremySkinner/FluentValidation#1006]
### System Details

- FluentValidation version: 8.1.2
- Web Framework version: ASP.NET Core 2.2

### Issue Description

We are trying to bundle up same rules for checking into one single Validator and use `SetValidator` to set to multiple difference query models.

However, it seems that when the property is null, validator will not be triggered. 

I had tried override `ValidateAsync` and `EnsureInstanceNotNull` as [Stackoverflow](https://stackoverflow.com/questions/17095791/fluentvalidation-rule-for-null-object/17095870) and https://github.com/JeremySkinner/FluentValidation/issues/486#issuecomment-308075903 suggested and has set `ImplicitlyValidateChildProperties = true`, but they are all not working.
If the parameter is not null, `ValidateAsync` and `EnsureInstanceNotNull` will trigger properly.
Bbut if the parameter is null, only the constructor is triggered while `ValidateAsync` and `EnsureInstanceNotNull` do not.

Here is a bare bone project construct with `FluentValidation` and `MediatR` to mimic our project setup
https://github.com/jasonycw/TestFluentValidation

Endpoint that is not performed as expected : `GET api/values/test4`

[/Controllers/Get.cs](https://github.com/jasonycw/TestFluentValidation/blob/master/TestFluentValidation/Controllers/Get.cs)
``` C#
public class Validator : AbstractValidator<Query>
        {
            public Validator()
                => RuleFor(x => x.Email)
                    //.NotNull()
                    //.WithState(x => new Error("Get.Validator", "Email is null"))
                    .SetValidator(new EmailValidator(false));
        }
```

If I put the NotNull() in outside of the `EmailValidator`, null check will be performed properly

**Comment by JeremySkinner** [comment]
Yes this is the normal/correct behaviour. Child validators invoked with SetValidator cannot be invoked if the property is null. You should put a NotNull check before the call to SetValidator.

**Comment by jasonycw** [comment]
Is there any suggestion on how to group them together?
Instead of `RuleFor(x => x.Something).NotNull().WithState(...).SetValidator(...)` repeat in every endpoint, is there a way to group the `NotNull` rule?

**Comment by JeremySkinner** [comment]
Wrap them both in an extension method?

**Comment by jasonycw** [comment]
But if the validation is use for many request object, the extension method will have no way to get the request property for error result handling
``` C#
public static void EmailAddress<T>(this IRuleBuilder<T, string> rule, bool acceptGmail)
{
    rule.Must(x => Email.Create(x) != null)
        .WithState(x => Errors.InvalidEmail(/** x.Email **/));
    rule.Must(x => Email.Create(x)?.IsGmail ?? false)
        .When(x => !acceptGmail)
        .WithState(x => Errors.InvalidEmail(/** x.Email **/));
}
```

FYI, we are trying to set 3 rules for request with email (string)
``` C#
//email string is not null
RuleFor(x => x.Email)
	.NotNull()
	.WithState(x => Errors.Null("Email"))` 
//email is valid
RuleFor(x => Email.Create(x.Email))
	.NotNull()
	.WithState(x => Errors.InvalidEmail(x.Email))` 
//(optional) email is not allowed
RuleFor(x => Email.Create(x.Email))
	.Must(x => !(x?.IsGmail ?? false))
	.When(x => !acceptGmail)
	.WithState(x => Errors.NotAllowed(x.Email)); 
```
These logic are duplicated a lot for many requests so we want to group them and reuse for difference reqeusts

**Comment by JeremySkinner** [comment]
So I was thinking more along the lines of something generic like a `SetValidatorWithNotNullCheck` extension method. You’d still use your original child validator but the extension method would do both the call to NotNull and SetValidator.

**Comment by JeremySkinner** [comment]
The other option would be to create an extension for the validator class itself, and put the rules in there:

```csharp
public static void EmailRules<T>(this AbstractValidator<T> validator, Expression<Func<T, string>> expression, bool acceptGmail) {
	var accessor = expression.Compile();
	
	validator.RuleFor(expression)
		.NotNull()
		.WithState(x => Errors.Null("Email"));

	validator.RuleFor(x => Email.Create(accessor(x)))
		.NotNull()
		.WithState(x => Errors.InvalidEmail(accessor(x)));
	
	validator.RuleFor(x => Email.Create(accessor(x)))
		.Must(x => !(x?.IsGmail ?? false))
		.When(x => !acceptGmail)
		.WithState(x => Errors.NotAllowed(accessor(x))); 
}
```

By taking the expression manually, you can compile it to a delegate which can be used to retrieve the property value (without needing to know the name of the property).

Then inside your main validator you can just do:

```csharp
public MyValidator() {
  this.EmailRules(x => x.Email);
}
```
